SoObjects/Mailer/SOGoTnefMailBodyPart.m

```diff
@@ -0,0 +1,955 @@
+    case 0x40:
+      switch (b) {
+        case 0xFA:
+          // Apr
+          return (4);
+      }
+      break;
+    case 0x20:
+      if (b == 0xFA) {
+        // Jul
+        return (7);
+      }
+      break;
+    case 0x80:
+      if (b == 0xA8) {
+        // Aug
+        return (8);
+      }
+      break;
+    case 0xA0:
+      if (b == 0xFF) {
+        // Oct
+        return (10);
+      }
+      break;
+    case 0xC0:
+      if (b == 0x56) {
+        return (12);
+      }
+  }
+
+  // Error
+  return (0);
+}
+
+@implementation SOGoTnefMailBodyPart
+
+/* Overwritten methods */
+
+- (id) init
+{
+  if ((self = [super init]))
+    {
+      debugOn = [[SOGoSystemDefaults sharedSystemDefaults] tnefDecoderDebugEnabled];
+      part = nil;
+      filename = nil;
+      bodyParts = [[NGMimeMultipartBody alloc] init];
+      [bodyParts retain];
+    }
+
+  return self;
+}
+
+- (id) initWithName: (NSString *) _name
+        inContainer: (id) _container
+{
+  self = [super initWithName: _name inContainer: _container];
+
+  [self decodeBLOB];
+
+  return self;
+}
+
+- (void) dealloc
+{
+  [part release];
+  [filename release];
+  [bodyParts release];
+
+  [super dealloc];
+}
+
+- (id) lookupName: (NSString *) _key
+	inContext: (id) _ctx
+	  acquire: (BOOL) _flag
+{
+  NSArray *parts;
+  int i;
+
+  if ([self isBodyPartKey: _key])
+    {
+      // _key is an integer
+      parts = [bodyParts parts];
+      i = [_key intValue] - 1;
+
+      if (i > -1 && i < [parts count])
+        {
+          [self setPart: [parts objectAtIndex: i]];
+          return self;
+        }
+    }
+  else if ([_key isEqualToString: [self filename]])
+    {
+      return self;
+    }
+  else if ([_key isEqualToString: @"asAttachment"])
+    {
+      [self setAsAttachment];
+      return self;
+    }
+
+  /* Fallback to super class */
+  return [super lookupName: _key inContext: _ctx acquire: _flag];
+}
+
+- (NSData *) fetchBLOB
+{
+  if (part)
+    return [part body];
+
+  return [super fetchBLOB];
+}
+
+- (NSString *) filename
+{
+  if (filename)
+    return filename;
+  else if (part)
+    return nil; // don't try to fetch the filename from the IMAP body structure
+  else
+    return [super filename];
+}
+
+- (id) partInfo
+{
+  if (partInfo)
+    return partInfo;
+  else if (part)
+    return nil; // don't try to fetch the info from the IMAP body structure
+  else
+    return [super partInfo];
+}
+
+/* New methods */
+
+- (NGMimeMultipartBody *) bodyParts
+{
+  return bodyParts;
+}
+
+- (void) decodeBLOB
+{
+  NSData *data;
+  NSEnumerator *list;
+  NSString *messageClass;
+  NSString *partName, *type, *subtype;
+  NSString *value, *attendee;
+  RTFHandler *handler;
+
+  DWORD signature;
+  DDWORD *classification;
+  dtr datetime;
+  TNEFStruct tnef;
+  variableLength *attachmentName;
+  variableLength *filedata;
+  variableLength buf;
+
+  [self setPart: nil];
+  partName = nil;
+  data = [self fetchBLOB];
+
+  memcpy(&signature, [data bytes], sizeof(DWORD));
```

Comment:
Make sure you free up signature - everywhere you do a memcpy, I don't think memory is freed.